Strengthened checking logic

aeharvlee opened the pull request:
## Proposed changes

* Check the whether NodeApi is initalized or not in WalletApi instance.
* WalletApi needs NodeApi when do migrateAccounts.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/ground-x/caver-java-ext-kas/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/ground-x/caver-java-ext-kas)
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

aeharvlee committed:
Strengthen checking logic

If the rpc endpoint url is set as localhost,
we consider this user does not initialize their Node API endpoint.

sirano11 commented on src/main/java/xyz/groundx/caver_ext_kas/kas/wallet/Wallet.java:
```diff
@@ -141,6 +141,12 @@ public RegistrationStatusResponse migrateAccounts(List<MigrationAccount> account
         if (this.rpc == null) {
             throw new NoSuchFieldException("Before using migrateAccounts, rpc must be set. You should call initNodeAPI first.");
         }
+        if (this.rpc.getWeb3jService() instanceof HttpService) {
```
Don't we need to consider condition if the Node API has initialized with websocket?

aeharvlee replied:
It would be great to add that condition too 🙂 .
I'll add the condition for `WebsocketService`. Thanks for your feedback.

aeharvlee committed:
Add endpoint url field in Wallet API

Because WebsocketService does not provide getter function for URL it contains,
we should track it as separate field.

sirano11 commented on src/main/java/xyz/groundx/caver_ext_kas/CaverExtKAS.java:
```diff
@@ -388,6 +388,7 @@ public void initWalletAPI(int chainId, String accessKeyId, String secretAccessKe
     public void initWalletAPI(String chainId, String accessKeyId, String secretAccessKey, String url) {
         kas.initWalletAPI(chainId, accessKeyId, secretAccessKey, url);
         kas.wallet.setRPC(this.rpc);
+        kas.wallet.setEndpointUrl(url);
```
I think it is incorrect code. Did you test this code?

aeharvlee replied:
Sorry for this. I fixed this logic.

sirano11 commented on src/main/java/xyz/groundx/caver_ext_kas/kas/wallet/Wallet.java:
```diff
@@ -107,6 +106,11 @@
+    /**
+     * The endpoint url set in RPC.
+     */
+    String endpointUrl;
```
What does mean endpointUrl? Wallet API's url? or Node API's url?

aeharvlee replied:
I changed variable name to rpcUrl.

aeharvlee committed:
Fix wrong logic and change variable name

sirano11 commented on src/main/java/xyz/groundx/caver_ext_kas/kas/wallet/Wallet.java:
```diff
@@ -141,6 +146,9 @@ public RegistrationStatusResponse migrateAccounts(List<MigrationAccount> account
         if (this.rpc == null) {
```
I think it can check a condition using a rpcUrl.

aeharvlee replied:
I agreed with that. 
I'll remove checking logic whether rpc is null or not.

aeharvlee committed:
Removed un-necessary logic

rpc field will not be null if the user created wallet api instance by new CaverExtKAS().

aeharvlee committed:
Add boolean to check node api is initialized or not

This commit also fixes
* typo at variable name
* test codes

aeharvlee committed:
Fix typo

sirano11 commented on src/main/java/xyz/groundx/caver_ext_kas/CaverExtKAS.java:
```diff
@@ -54,6 +54,11 @@
+    /**
+     * The boolean to check whether NodeApi is initialized or not.
+     */
+    public boolean nodeApiInitialized;
```
Does this field need to set a public?

sirano11 commented on src/main/java/xyz/groundx/caver_ext_kas/wallet/KASWallet.java:
```diff
@@ -55,7 +55,18 @@
+     * @param walletAPI An WalletAPI instance to use KAS Wallet API.
+     * @param nodeApiInitialized A boolean to check whether the NodeApi is initialized or not.
+     */
+    public KASWallet(Wallet walletAPI, boolean nodeApiInitialized) {
```
Why does need this constructor?

aeharvlee replied:
In most case, I think it will be used for internal purposes only.
So I think it should be private.

Thanks for your feedback. I'll update it.

sirano11 commented on src/main/java/xyz/groundx/caver_ext_kas/CaverExtKAS.java:
```diff
@@ -388,7 +397,7 @@ public void initWalletAPI(int chainId, String accessKeyId, String secretAccessKe
         kas.initWalletAPI(chainId, accessKeyId, secretAccessKey, url);
         kas.wallet.setRPC(this.rpc);
-        setWallet(new KASWallet(this.kas.wallet));
+        setWallet(new KASWallet(this.kas.wallet, nodeApiInitialized));
```
Don't we need to set a nodeAPIInitialized? in kas.wallet instance?

aeharvlee committed:
Refactored codes

1. Change variable name
2. Make it private
3. Update the logic correctly.
